Mathlib/Data/PFunctor/Multivariate/W.lean

```diff
@@ -0,0 +1,319 @@
+    (g' : P.drop.B a ⟹ α) (g : ∀ j : P.last.B a, P.WPath (f j) ⟹ α) :
+    P.wPathDestRight (P.wPathCasesOn g' g) = g := rfl
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.W_path_dest_right_W_path_cases_on MvPFunctor.wPathDestRight_wPathCasesOn
+
+theorem wPathCasesOn_eta {α : TypeVec n} {a : P.A} {f : P.last.B a → P.last.W}
+    (h : P.WPath ⟨a, f⟩ ⟹ α) : P.wPathCasesOn (P.wPathDestLeft h) (P.wPathDestRight h) = h := by
+  ext (i x) ; cases x <;> rfl
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.W_path_cases_on_eta MvPFunctor.wPathCasesOn_eta
+
+theorem comp_wPathCasesOn {α β : TypeVec n} (h : α ⟹ β) {a : P.A} {f : P.last.B a → P.last.W}
+    (g' : P.drop.B a ⟹ α) (g : ∀ j : P.last.B a, P.WPath (f j) ⟹ α) :
+    h ⊚ P.wPathCasesOn g' g = P.wPathCasesOn (h ⊚ g') fun i => h ⊚ g i := by
+  ext (i x) ; cases x <;> rfl
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.comp_W_path_cases_on MvPFunctor.comp_wPathCasesOn
+
+/-- Polynomial functor for the W-type of `P`. `A` is a data-less well-founded
+tree whereas, for a given `a : A`, `B a` is a valid path in tree `a` so
+that `Wp.obj α` is made of a tree and a function from its valid paths to
+the values it contains  -/
+def wp : MvPFunctor n where
+  A := P.last.W
+  B := P.WPath
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.Wp MvPFunctor.wp
+
+/-- W-type of `P` -/
+-- Porting note: used to have @[nolint has_nonempty_instance]
+def W (α : TypeVec n) : Type _ :=
+  P.wp.Obj α
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.W MvPFunctor.W
+
+instance mvfunctorW : MvFunctor P.W := by delta MvPFunctor.W ; infer_instance
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.mvfunctor_W MvPFunctor.mvfunctorW
+
+/-!
+First, describe operations on `W` as a polynomial functor.
+-/
+
+
+/-- Constructor for `wp` -/
+def wpMk {α : TypeVec n} (a : P.A) (f : P.last.B a → P.last.W) (f' : P.WPath ⟨a, f⟩ ⟹ α) :
+    P.W α :=
+  ⟨⟨a, f⟩, f'⟩
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.Wp_mk MvPFunctor.wpMk
+
+def wpRec {α : TypeVec n} {C : Type _}
+    (g : ∀ (a : P.A) (f : P.last.B a → P.last.W), P.WPath ⟨a, f⟩ ⟹ α → (P.last.B a → C) → C) :
+    ∀ (x : P.last.W) (_ : P.WPath x ⟹ α), C
+  | ⟨a, f⟩, f' => g a f f' fun i => wpRec g (f i) (P.wPathDestRight f' i)
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.Wp_rec MvPFunctor.wpRec
+
+theorem wpRec_eq {α : TypeVec n} {C : Type _}
+    (g : ∀ (a : P.A) (f : P.last.B a → P.last.W), P.WPath ⟨a, f⟩ ⟹ α → (P.last.B a → C) → C)
+    (a : P.A) (f : P.last.B a → P.last.W) (f' : P.WPath ⟨a, f⟩ ⟹ α) :
+    P.wpRec g ⟨a, f⟩ f' = g a f f' fun i => P.wpRec g (f i) (P.wPathDestRight f' i) := rfl
+set_option linter.uppercaseLean3 false in
+#align mvpfunctor.Wp_rec_eq MvPFunctor.wpRec_eq
+
+-- Note: we could replace Prop by Type _ and obtain a dependent recursor
+theorem Wp_ind {α : TypeVec n} {C : ∀ x : P.last.W, P.WPath x ⟹ α → Prop}
```

Comment:
Should this be `wp_ind`, to be consistent with the capitalization of `wp`?

Reply:
Yes. I'll fix it when the dependency is merged.